src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj

```diff
@@ -196,10 +199,24 @@
    timeout :- schema/Int]
   {:pre  [(>= timeout 0)]}
   (borrow-from-pool!* (partial borrow-with-timeout-fn timeout)
-                      (get-pool pool-context) pool-context))
+                      (get-pool pool-context)))
 
 (schema/defn ^:always-validate
   return-to-pool
   "Return a borrowed pool instance to its free pool."
   [instance :- jruby-schemas/JRubyPuppetInstanceOrRetry]
-  (.putFirst (:pool instance) instance))
+  (if (jruby-schemas/jruby-puppet-instance? instance)
+    (let [new-state (swap! (:state instance)
+                           update-in [:request-count] inc)
+          {:keys [max-requests flush-instance-fn pool]} instance]
+      (if (and (pos? max-requests)
```

Comment:
It looks like `max-requests` can only be a positive integer or `nil` here ... if that's the case, I think it's better to use `(not (nil? ...` instead of `pos?` here.